v-next/hardhat/src/internal/builtin-plugins/solidity/build-results.ts

```diff
@@ -109,46 +98,75 @@ export async function getBuildInfos(
 }
 
 /**
- * This function returns the artifacts generated during the compilation associated
- * with the given Solidity build results. It relies on the fact that each successful
- * build result has a corresponding artifact generated property.
+ * This function returns the artifacts generated during the compilation.
  *
- * @param results The successful Solidity build results.
+ * @param artifactManager The artifact manager.
  * @returns The artifacts in the format expected by the EDR.
  */
 export async function getArtifacts(
-  results: SuccessfulSolidityBuildResults,
+  artifactManager: ArtifactManager,
 ): Promise<EdrArtifact[]> {
-  const contractArtifacts = Array.from(results.entries())
-    .map(([source, result]) => {
-      return result.contractArtifactsGenerated.map((artifactPath) => ({
-        source,
-        solcVersion: result.compilationJob.solcConfig.version,
-        artifactPath,
-      }));
-    })
-    .flat();
+  const fullyQualifiedNames = await artifactManager.getAllFullyQualifiedNames();
 
-  return Promise.all(
-    contractArtifacts.map(async ({ source, artifactPath, solcVersion }) => {
-      const artifact: HardhatArtifact = await readJsonFile(artifactPath);
+  const artifacts = await Promise.all(
+    Array.from(fullyQualifiedNames).map(async (fullyQualifiedName) => {
+      return artifactManager.readArtifact(fullyQualifiedName);
+    }),
+  );
 
-      const id = {
-        name: artifact.contractName,
-        solcVersion,
-        source,
-      };
+  const buildInfoIds = Array.from(
+    new Set(artifacts.map((artifact) => artifact.buildInfoId)),
+  );
 
-      const contract = {
-        abi: JSON.stringify(artifact.abi),
-        bytecode: artifact.bytecode,
-        deployedBytecode: artifact.deployedBytecode,
-      };
+  const solcVersionsArray: Array<[string, string]> = await Promise.all(
+    buildInfoIds.map(async (buildInfoId) => {
+      assertHardhatInvariant(
+        buildInfoId !== undefined,
+        "artifactBuildInfoId should not be undefined",
+      );
 
-      return {
-        id,
-        contract,
-      };
+      const buildInfoPath = await artifactManager.getBuildInfoPath(buildInfoId);
+
+      assertHardhatInvariant(
+        buildInfoPath !== undefined,
+        "buildInfoPath should not be undefined",
+      );
+
+      const buildInfo: BuildInfo = await readJsonFile(buildInfoPath);
+
+      return [buildInfoId, buildInfo.solcVersion];
     }),
   );
+  const solcVersions = new Map(solcVersionsArray);
+
+  return artifacts.map((artifact) => {
+    assertHardhatInvariant(
+      artifact.buildInfoId !== undefined,
+      "solcVersion should not be undefined",
+    );
+
+    const solcVersion = solcVersions.get(artifact.buildInfoId);
+
+    assertHardhatInvariant(
+      solcVersion !== undefined,
+      "solcVersion should not be undefined",
+    );
+
+    const id = {
+      name: artifact.contractName,
+      solcVersion,
+      source: artifact.sourceName,
```

Comment:
I wonder what value should we use for npm artifacts here?

Reply:
It should be the `artifact.inputSourceName`, otherwise EDR won't find it.

We should polish the `sourceName` vs `inputSourceName` vs `publicSourceName` situation.

Reply:
but for now, something like `artifact.inputSourceName ?? artifact.sourceName`.